src/Helpers/Intent.sol

```diff
@@ -0,0 +1,108 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.17;
+
+import { LibClone } from "solady/utils/LibClone.sol";
+import { SafeTransferLib } from "solady/utils/SafeTransferLib.sol";
+import { IIntent } from "../Interfaces/IIntent.sol";
+
+interface IERC20 {
+    function balanceOf(address account) external view returns (uint256);
+}
+
+/// @title Intent
+/// @author LI.FI (https://li.fi)
+/// @notice Intent contract that can execute arbitrary calls.
+/// @custom:version 1.0.0
+contract Intent {
+    /// Storage ///
+
+    bytes32 public intentId;
+    bytes32 public salt;
+    address public receiver;
+    address public immutable implementation;
+    address public factory;
+    address public tokenOut;
+    uint256 public amountOutMin;
+    bool public executed = false;
+
+    /// Errors ///
+
+    error Unauthorized();
+    error AlreadyExecuted();
+    error InvalidParams();
+    error ExecutionFailed();
+    error InsufficientOutputAmount();
+
+    /// Constructor ///
+
+    constructor() {
+        implementation = address(this);
+    }
+
+    /// External Methods ///
+
+    /// @notice Initializes the intent with the given parameters.
+    /// @param _initData The init data.
+    function init(IIntent.InitData calldata _initData) external {
+        salt = keccak256(abi.encode(_initData));
+        factory = msg.sender;
+        address predictedAddress = LibClone.predictDeterministicAddress(
+            implementation,
+            salt,
+            msg.sender
+        );
+        if (address(this) != predictedAddress) {
+            revert InvalidParams();
+        }
+
+        intentId = _initData.intentId;
+        receiver = _initData.receiver;
+        tokenOut = _initData.tokenOut;
+        amountOutMin = _initData.amountOutMin;
+    }
+
+    /// @notice Executes the intent with the given calls.
+    /// @param calls The calls to execute.
+    function execute(IIntent.Call[] calldata calls) external {
+        if (msg.sender != factory) {
+            revert Unauthorized();
+        }
+        if (executed) {
+            revert AlreadyExecuted();
+        }
+        executed = true;
+
+        for (uint256 i = 0; i < calls.length; i++) {
+            (bool success, ) = calls[i].to.call{ value: calls[i].value }(
+                calls[i].data
+            );
+            if (!success) {
+                revert ExecutionFailed();
+            }
+        }
+
+        if (IERC20(tokenOut).balanceOf(address(this)) < amountOutMin) {
+            revert InsufficientOutputAmount();
+        }
+        if (tokenOut == address(0)) {
```

Comment:
The issue doesn't actually apply here as both Metis and Mode still have the ability to move ETH around in a native manner and balances still work the same. The issue with those chains stemmed from the fact that different exchanges were using the precompile addresses for native and treating native as an ERC20 instead of native which conflicted with the checks we have in our swap logic.

Reply:
ok